src/Microsoft.AspNet.Mvc.Core/ModelBinding/DictionaryModelBinder.cs

```diff
@@ -14,14 +17,86 @@ namespace Microsoft.AspNet.Mvc.ModelBinding
     public class DictionaryModelBinder<TKey, TValue> : CollectionModelBinder<KeyValuePair<TKey, TValue>>
     {
         /// <inheritdoc />
+        public override async Task<ModelBindingResult> BindModelAsync([NotNull] ModelBindingContext bindingContext)
+        {
+            var result = await base.BindModelAsync(bindingContext);
+            if (result == null || !result.IsModelSet)
+            {
+                // No match for the prefix at all.
+                return result;
+            }
+
+            var model = result.Model as Dictionary<TKey, TValue>;
+            Debug.Assert(model != null);
+            if (model.Count != 0)
+            {
+                // ICollection<KeyValuePair<TKey, TValue>> approach was successful.
+                return result;
+            }
+
+            var enumerableValueProvider = bindingContext.ValueProvider as IEnumerableValueProvider;
```

Comment:
So, practically speaking then - you could hit this case if you put `[FromRoute]` on your model? It would be good to document what reasonable cases lead to this for clarity, that's my only concern.